jupyterhub_traefik_proxy/performance/proxy_perf.py

```diff
@@ -0,0 +1,294 @@
+        "--measure",
+        dest="metric",
+        default="methods",
+        help=textwrap.dedent(
+            """\
+            What metric to measure. Available metrics:
+            - methods
+            - http_throughput_small
+            - http_throughput_large
+            - ws_throughput
+            If no metric is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--proxy",
+        dest="proxy_class",
+        default="TomlProxy",
+        help=textwrap.dedent(
+            """\
+            Proxy class to analyze.
+            Available proxies:
+            -TomlProxy
+            -EtcdProxy
+            -CHP
+            If no proxy is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--routes_number",
+        dest="routes_number",
+        default=10,
+        help=textwrap.dedent(
+            """\
+            Number of routes to be added/deleted/retrieved.
+            If no number is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--concurrent_requests_number",
+        dest="concurrent_requests_number",
+        default=10,
+        help=textwrap.dedent(
+            """\
+            Number of concurrent requests when comoputing the throughput.
+            If no number is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--output",
+        dest="csv_filename",
+        help=textwrap.dedent(
+            """
+            The csv file name where the results will be stored.
+            If no file is provided, the results will only be outputed to stdout.
+            """
+        ),
+    )
+
+    parser.add_argument(
+        "--backend_port",
+        dest="backend_port",
+        default=9000,
+        help=textwrap.dedent(
+            """
+            The port where the backend receives http/websocket requests.
+            If no port is provided, it defaults to:
+            --- %(default)s ---
+            """
+        ),
+    )
+
+    args = parser.parse_args()
+    metric = args.metric
+    proxy_class = args.proxy_class
+    routes_number = int(args.routes_number)
+    concurrent_requests_number = int(args.concurrent_requests_number)
+    csv_filename = args.csv_filename
+    backend_port = args.backend_port
+
+    requests_no = 1000
+
+    loop = asyncio.get_event_loop()
+    if metric == "methods":
+        result = loop.run_until_complete(
+            measure_methods_performance(proxy_class, routes_number)
+        )
+        if csv_filename:
+            with open(csv_filename, mode="w+") as csv_file:
+                fieldnames = [("T_" + str(i)) for i in range(routes_number)]
+                fieldnames.insert(0, "Method")
+                fieldnames.append("Mean")
```

Comment:
Mean doesn't need to be recorded, since it can be easily computed from the samples. Keeping the numbers in a given row or column meaning the same thing helps with plotting/summary statistics later.